### PR TITLE
crimson, common: introduce ceph::atomic and apply it on bufferlist.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by mattbenjamin** [comment]
interesting, seems a very clean take on non-atomic atomics

File: src/common/ceph_atomic.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    std::enable_if_t<std::is_integral_v<T>, T> operator++(int) {
+      return value++;
+    }
+    std::enable_if_t<std::is_integral_v<T>, T> operator--() {
+      return --value;
+    }
+    std::enable_if_t<std::is_integral_v<T>, T> operator--(int) {
+      return value--;
+    }
+    std::enable_if_t<std::is_integral_v<T>, T> operator+=(const dummy_atomic& b) {
```
Comment (tchaikov) [inline src/common/ceph_atomic.h]: yeah, i wonder if we should allow floating-point types in future once we move to C++20.

**Comment by tchaikov** [comment]
```
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 11.1888/12.2818  => rejected
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 2864.0/3144.0  => rejected
22:05:16 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 219.42/160.83  => accepted
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00557787/0.00502689  => rejected
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 11.1896/12.1108  => rejected
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 2864.0/3100.0  => rejected
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 231.065/140.353  => rejected
22:05:16 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.0055844/0.00510875  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 48.1724/53.0009  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 12332.0/13568.0  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 1284.24/162.681  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.0012906/0.00117155  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 48.4246/52.9538  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 12396.0/13556.0  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 1240.63/160.461  => rejected
22:05:16 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00128396/0.00117218  => rejected
22:05:16 - WARNING  - cbt      - 15 tests failed out of 16
```

the CBT test result does not make sense anymore.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-01-25_04:55:56-rados-wip-kefu-testing-2020-01-25-0051-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-01-24_02:23:44-rados-wip-kefu-testing-2020-01-23-2212-distro-basic-smithi/

failures tracked by 

- https://tracker.ceph.com/issues/43703
- https://tracker.ceph.com/issues/43653
- https://tracker.ceph.com/issues/36353

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by tchaikov** [comment]
```
0.05)):: 12.2803/11.0471  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 3143.0/2828.0  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 189.391/150.998  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00503416/0.00564253  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 12.1895/11.1203  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 3120.0/2846.0  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 151.583/174.277  => accepted
04:08:39 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00508801/0.00560708  => accepted
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 49.3061/51.0889  => accepted
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 12622.0/13078.0  => accepted
04:08:39 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 0.50)):: 1196.85/209.023  => rejected
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00125977/0.00121642  => accepted
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 49.0258/50.7566  => accepted
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 12550.0/12993.0  => accepted
04:08:39 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 0.50)):: 1196.92/170.338  => rejected
04:08:39 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00126806/0.00122478  => accepted
04:08:39 - WARNING  - cbt      - 2 tests failed out of 16
```

**Pull request merged**